File: code/packages/python/network-stack/src/network_stack/udp.py
"""
UDP — Layer 4 (Transport), Unreliable Datagrams
================================================

UDP (User Datagram Protocol) is the simpler sibling of TCP. Where TCP provides
reliable, ordered, connection-oriented byte streams, UDP provides unreliable,
unordered, connectionless **datagrams**.

Why Use UDP?
------------

If TCP is registered mail with tracking, UDP is a postcard. Postcards are:

- **Faster**: No three-way handshake, no acknowledgments, no retransmissions.
  You write the message and drop it in the mailbox.
- **Simpler**: No state machine, no sequence numbers, no flow control.
- **Unreliable**: The postcard might get lost, arrive out of order, or arrive
  twice. The sender never knows.

UDP is used when speed matters more than reliability:

- **DNS**: A single question-and-answer exchange. If the reply is lost,
  just ask again.
- **Video streaming**: A lost frame is better than a delayed frame.
- **Online gaming**: The newest position update matters; old ones are
  irrelevant.

UDP Header Format
-----------------

The UDP header is beautifully simple — just 8 bytes::

    0                   1                   2                   3
    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |          Source Port          |       Destination Port        |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |            Length             |           Checksum            |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+

Compare this with TCP's 20 bytes of header overhead. UDP trades features
for efficiency.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass


@dataclass
class UDPHeader:
    """
    A UDP datagram header — exactly 8 bytes.

    Attributes
    ----------
    src_port : int
        Source port (0-65535). Can be 0 if the sender doesn't expect a reply.
    dst_port : int
        Destination port (0-65535).
    length : int
        Total datagram size (header + payload) in bytes. Minimum 8 (header
        only, no payload).
    checksum : int
        Error detection. Can be 0 (meaning "no checksum computed") in IPv4.
        We always set it to 0 in this teaching implementation.
    """

    src_port: int = 0
    dst_port: int = 0
    length: int = 8
    checksum: int = 0

    def serialize(self) -> bytes:
        """
        Convert the header to 8 raw bytes in network byte order.

        Layout::

            Bytes 0-1: src_port
            Bytes 2-3: dst_port
            Bytes 4-5: length
            Bytes 6-7: checksum
        """
        return struct.pack(
            "!HHHH",
            self.src_port,
            self.dst_port,
            self.length,
            self.checksum,
        )

    @classmethod
    def deserialize(cls, data: bytes) -> UDPHeader:
        """
        Parse 8 bytes into a UDPHeader.

        Raises ValueError if data is too short.
        """
        if len(data) < 8:
            msg = f"UDP header too short: {len(data)} bytes (minimum 8)"
            raise ValueError(msg)

        src_port, dst_port, length, checksum = struct.unpack("!HHHH", data[:8])

        return cls(
            src_port=src_port,
            dst_port=dst_port,
            length=length,
            checksum=checksum,
        )


class UDPSocket:
    """
    A UDP socket — sends and receives individual datagrams.

    Unlike TCP, there is no connection setup. Each datagram is independent:
    you specify the destination address with every send, and each received
    datagram tells you where it came from.

    Think of this like a mailbox: you can send a letter to anyone without
    establishing a relationship first, and you can receive letters from
    anyone who knows your address.

    Parameters
    ----------
    local_port : int
        The port this socket is bound to. Other hosts send to this port.

    Example
    -------
    >>> sock = UDPSocket(local_port=53)
    >>> header, payload = sock.send_to(b"query", dest_ip=0x08080808, dest_port=53)
    >>> sock.deliver(b"response", src_ip=0x08080808, src_port=53)
    >>> data, ip, port = sock.receive_from()
    """

    def __init__(self, local_port: int = 0) -> None:
        self.local_port = local_port
        # Queue of received datagrams: (data, src_ip, src_port)
        self._recv_queue: list[tuple[bytes, int, int]] = []

    def send_to(
        self, data: bytes, dest_ip: int, dest_port: int
    ) -> tuple[UDPHeader, bytes]:
        """
        Create a UDP datagram for the given destination.

        Returns the header and payload bytes. The caller (IP layer) is
        responsible for actually transmitting the datagram.

        Parameters
        ----------
        data : bytes
            The payload to send.
        dest_ip : int
            Destination IP (not used in the header, but returned for
            the caller's convenience).
        dest_port : int
            Destination port number.

        Returns
        -------
        tuple[UDPHeader, bytes]
            The UDP header and the payload.
        """
        header = UDPHeader(
            src_port=self.local_port,
            dst_port=dest_port,
            length=8 + len(data),
        )
        return header, data

    def receive_from(self) -> tuple[bytes, int, int] | None:
        """
        Receive the next datagram from the queue.

        Returns (data, src_ip, src_port) or None if no datagrams are
        available.

        Unlike TCP's ``recv()``, this returns the sender's address because
        UDP is connectionless — each datagram can come from a different host.
        """
        if not self._recv_queue:
            return None
        return self._recv_queue.pop(0)

    def deliver(self, data: bytes, src_ip: int, src_port: int) -> None:
        """
        Deliver a received datagram to this socket.

        This simulates the kernel's job of demultiplexing incoming datagrams
        and routing them to the correct socket based on the destination port.

        Parameters
        ----------
        data : bytes
            The datagram payload.
        src_ip : int
            Source IP address (who sent this).
        src_port : int
            Source port number.
        """
        self._recv_queue.append((data, src_ip, src_port))

    @property
    def has_data(self) -> bool:
        """Return True if there are datagrams waiting to be read."""
        return len(self._recv_queue) > 0

File: code/packages/ruby/network_stack/lib/coding_adventures/network_stack/udp.rb
# frozen_string_literal: true

# ============================================================================
# Layer 4: UDP — User Datagram Protocol
# ============================================================================
#
# UDP is the "anti-TCP" — it provides no reliability, no ordering, no flow
# control, and no connection state. You send a datagram (a self-contained
# packet of data); it either arrives or it doesn't. There are no
# acknowledgments, no retransmissions, and no sequence numbers.
#
# Why would anyone use such a "unreliable" protocol? Because simplicity
# and speed matter for certain applications:
#
#   - DNS lookups: one question, one answer — if it's lost, just ask again.
#   - Video streaming: a dropped frame is better than pausing to retransmit.
#   - Online games: the latest position update matters; old ones are useless.
#   - VoIP: stuttering is better than silence followed by a burst of audio.
#
# The entire UDP header is only 8 bytes (vs. TCP's minimum 20). This means
# less overhead per packet, which adds up when you're sending thousands of
# small messages per second.
#
# UDP Header format:
#
#    0                   1                   2                   3
#    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
#   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
#   |          Source Port          |       Destination Port        |
#   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
#   |            Length             |           Checksum            |
#   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
#
# ============================================================================

module CodingAdventures
  module NetworkStack
    # ========================================================================
    # UDPHeader
    # ========================================================================
    #
    # The 8-byte UDP header. Compare its simplicity to TCPHeader above —
    # no sequence numbers, no acknowledgments, no flags, no window size.
    # Just source port, destination port, length, and an optional checksum.
    #
    # ========================================================================
    class UDPHeader
      attr_accessor :src_port, :dst_port, :length, :checksum

      def initialize(src_port:, dst_port:, length: 8, checksum: 0)
        @src_port = src_port
        @dst_port = dst_port
        @length   = length
        @checksum = checksum
      end

      # Serialize the UDP header into an 8-byte array.
      def serialize
        bytes = Array.new(8, 0)
        bytes[0] = (@src_port >> 8) & 0xFF
        bytes[1] = @src_port & 0xFF
        bytes[2] = (@dst_port >> 8) & 0xFF
        bytes[3] = @dst_port & 0xFF
        bytes[4] = (@length >> 8) & 0xFF
        bytes[5] = @length & 0xFF
        bytes[6] = (@checksum >> 8) & 0xFF
        bytes[7] = @checksum & 0xFF
        bytes
      end

      # Deserialize an 8-byte array into a UDPHeader.
      def self.deserialize(bytes)
        return nil if bytes.length < 8

        src_port = (bytes[0] << 8) | bytes[1]
        dst_port = (bytes[2] << 8) | bytes[3]
        length   = (bytes[4] << 8) | bytes[5]
        checksum = (bytes[6] << 8) | bytes[7]

        new(src_port: src_port, dst_port: dst_port, length: length, checksum: checksum)
      end
    end

    # ========================================================================
    # UDPSocket — Connectionless Datagram Socket
    # ========================================================================
    #
    # Unlike TCP, a UDP socket has no connection state. You can send a
    # datagram to any address at any time (send_to), and datagrams from
    # any address arrive in a queue (receive_from).
    #
    # The recv_queue stores tuples of (data, src_ip, src_port) so the
    # application knows where each datagram came from — important because
    # there is no "connection" to track this automatically.
    #
    # ========================================================================
    class UDPSocket
      attr_reader :local_port, :recv_queue

      def initialize(local_port:)
        @local_port = local_port
        @recv_queue = []
      end

      # Prepare a datagram for sending. Returns [UDPHeader, data].
      #
      # Note: UDP doesn't actually "connect" to anything — each send_to
      # specifies the destination independently. The header just says
      # "from port X, to port Y, this many bytes."
      #
      def send_to(data, dst_port)
        header = UDPHeader.new(
          src_port: @local_port,
          dst_port: dst_port,
          length: 8 + data.length
        )
        [header, data]
      end

      # Deliver an incoming datagram to this socket.
      #
      # Called by the network stack when a UDP packet arrives addressed to
      # our port. The application will later call receive_from to retrieve it.
      #
      def deliver(data, src_ip, src_port)
        @recv_queue.push({data: data, src_ip: src_ip, src_port: src_port})
      end

      # Retrieve the next datagram from the receive queue.
      #
      # Returns {data:, src_ip:, src_port:} or nil if the queue is empty.
      #
      def receive_from
        @recv_queue.shift
      end
    end
  end
end
